File: JavaScript/Big-O/excercises.js
// What is the Big O of the below function?
function funChallenge(input) {
  let a = 10; // O(1)
  a = 50 + 3; // O(1)

  for (let i = 0; i < input.length; i++) {
    anotherFunction(); // O(n)
    let stranger = true; // O(n)
    a++; // O(n)
  }
  return a; // O(1)
}

// BIG O(3 + 4n) --> O(n)

// What is the Big O of the below function?
function anotherFunChallenge(input) {
  let a = 5; // O(1)
  let b = 10; // O(1)
  let c = 50; // O(1)
  for (let i = 0; i < input; i++) {
    let x = i + 1; // O(n)
    let y = i + 2; // O(n)
    let z = i + 3; // O(n)
  }
  for (let j = 0; j < input; j++) {
    let p = j * 2; // O(n)
    let q = j * 2; // O(n)
  }
  let whoAmI = "I don't know"; // O(1)
}

// BIG O(4 + 5n) --> O(n)

// Log all pairs of array
const boxes = ['a', 'b', 'c', 'd', 'e'];

function log_all_pairs_of_array(arr) {
  for (let i in boxes) {
    for (let j in boxes) {
      console.log(boxes[i], boxes[j]);
    }
  }
}

log_all_pairs_of_array(boxes);

// BIG O(n * n) --> O(n^2) (Quadratic time)

// Find 1st, find Nth...
const arr = [
  {
    tweet: 'hi',
    date: 2012,
  },
  {
    tweet: 'my',
    date: 2018,
  },
  {
    tweet: 'teddy',
    date: 2018,
  },
];  // O(n^2)

File: Python/Big-O/excercises.py
def randomFunc():
    return

# What is the Big O of the below function?


def funChallenge(input):
    a = 10  # O(1)
    a = 50 + 3  # O(1)
    for _ in range(len(input)):  # O(n)
        randomFunc()  # O(n)
        randomVar = True  # O(n)
    return a  # O(1)

# BIG O(3 + 4n) --> O(n)

# What is the Big O of the below function?


def anotherFunChallenge(input):
    a = 5  # O(1)
    b = 10  # O(1)
    c = 50  # O(1)
    for i in range(len(input)):
        x = i + 1  # O(n)
        y = i + 2  # O(n)
        z = i + 3  # O(n)

    for j in range(len(input)):
        p = j * 2  # O(n)
        q = j * 2  # O(n)

    whoAmI = "I don't know"  # O(1)

# BIG O(4 + 5n) --> O(n)


# Log all pairs of array
boxes = ['a', 'b', 'c', 'd', 'e']


def log_all_pairs_of_array(arr):
    for i in range(len(arr)):
        for j in range(len(arr)):
            print(i, j)


log_all_pairs_of_array(boxes)

# Find 1st, find Nth...
arr = [
    {
        "tweet": 'hi',
        "date": 2012,
    },
    {
        "tweet": 'my',
        "date": 2018,
    },
    {
        "tweet": 'teddy',
        "date": 2018,
    },
]  # O(n^2)

# Greatest number


def greatest_number(arr):   # O(n2)
    is_val_the_greatest = True
    for i in arr:
        for j in arr:
            if j > i:
                is_val_the_greatest = False
        if is_val_the_greatest:
            return i


def greatest_number_n2(arr):  # O(n)
    greatest_num = 0
    for i in arr:
        if i > greatest_num:
            greatest_num = i
    return greatest_num


print(greatest_number_n2([1, 5, 3, 2]))
